### PR TITLE
feat: allow backporting to remote repository

### PR DESCRIPTION
**Pull request by tasso94**
This pull request is still a work in progress, but I wanted to share my first iteration to get feedback early on:
1. Whether it generally makes sense and if you want to merge it with your upstream.
2. If the technical design of the feature makes sense.

## User story

As a user, I maintain my `master` branch and the backport branches in two separate repositories to hide my backport branches from the public.

## Open tasks

I think the following topics still need to be addressed (is there even more I might not know yet?):
* [x] Improve the naming of newly introduced config properties (I'm open to any ideas).
* [x] Write documentation.

### CODE REVIEW

**Comment by korthout** [comment]
Thanks @tasso94 🙇 I'm currently enjoying some time off, so it may take me some more time until I review this. But at first glance, this looks amazing!

File: action.yml
Comment (korthout) [inline action.yml:96]: 👍 I like the flexibility and naming. 

🔧 For some additional suggestions, we could consider `remote_owner` and `remote_repo` as valid options. These might overcome the potential naming confusion mentioned [here](https://github.com/korthout/backport-action/issues/384#issuecomment-1715746551). When they are not specified, the current working repo is used. Using `remote` in the naming clarifies what this really does: backport to a remote repo.

💭 We could also consider a single `target_repo` input that takes the full repo identifier `owner/name`. This would allow setting the default using the [github context](https://docs.github.com/en/actions/learn-github-actions/contexts#github-context) to `${{ github.repository }}`. Sadly the github context does not have the repo name only. A downside to this approach is that we must parse the string to separate the owner and repo for later usage. Additionally, if we have some logic specific to the remote target, then we must compare the value to the context's value to see if it was explicitly defined. So, I'm leaning away from the single `target_repo` input approach.

File: action.yml
Comment (korthout) [inline action.yml:86]: ❌ I think this change doesn't fully achieve what you intend.

This default was a template literal (or template string), where the variables `${pull_number}` and `${target}` are interpolated on execution. 

```ts
const branchname = `backport-${pull_number}-to-${target}`;
```

But now the `pull_branchname` is a special input where placeholders can be replaced with values. These are defined by utils:

https://github.com/korthout/backport-action/blob/930286d359d53effaf69607223933cbbb02460eb/src/utils.ts#L3-L21

Note that `${pull_number}` is a valid placeholder, but `${target}` is not. 

💡 Instead of `${target}`, you need to use `${target_branch}`.

File: src/git.ts
Comment (korthout) [inline src/git.ts:72]: ❓ Can you share where you got the `x-access-token:` prefix from? I can only find [other approaches](https://stackoverflow.com/a/66156992).

File: action.yml
Comment (korthout) [inline action.yml:96]: 🔧 As it may be difficult to consider all different use cases, I suggest we move the `target_owner` and `target_repo` inputs to the experimental input. This will help clarify to users that this functionality is available but should be tried out carefully before fully trusting it. Once it has been available for some time, we can promote it to a fully supported feature.

**Review by korthout — CHANGES_REQUESTED** [review]
Thanks for this massive contribution @tasso94 🙇 This will be super useful!

I've had a closer look at the changes and have a few comments, suggestions, and questions. I've used an [emoji review code](https://github.com/camunda/zeebe/blob/main/CONTRIBUTING.md#review-emoji-code) to clarify my comments. Please have a look.

**Comment by tasso94** [comment]
Hi @korthout,

Thanks for your review feedback, and I'm sorry for my late response.

Right now, I'm a bit busy but I'll try to have a look soon.

Stay tuned!

Best,
Tassilo

File: src/backport.ts
Comment (tasso94) [reply src/backport.ts]: Good point 👍 I changed it according to your proposal.

File: action.yml
Comment (tasso94) [reply action.yml]: Good point 👍 Changed it accordingly.

File: action.yml
Comment (tasso94) [reply action.yml]: Nice catch! 👍

File: src/git.ts
Comment (tasso94) [reply src/git.ts]: I removed passing the token to `git.ts`. Instead, the token can be passed to the checkout step.

```yml
- uses: actions/checkout@v4
  with:
    token: ${{ secrets.MY_TOKEN }}
```

File: action.yml
Comment (tasso94) [reply action.yml]: Moved it to the experimental input 👍

**Comment by tasso94** [comment]
Hi @korthout,

Thank you for providing your input. I applied your review hints. 🙂 👍 
Could you please review my changes?

I know that I still need to document the new properties. If you don't have any further remarks regarding the naming or characteristics of the properties, I will add the documentation.

Is there anything else that is required from my side to make this PR merge-ready? Do you want me to contribute tests?

Best,
Tassilo

**Comment by korthout** [comment]
Thanks for the great work @tasso94. I want to give this the proper attention it deserves, but I've been too busy at work as well as in my personal life (with an international move coming up ahead) to review this. Please know that I haven't forgotten about this. I really appreciate this effort and see it as a big improvement to the action. A full review will follow.

**Comment by tasso94** [comment]
Hi @korthout,

Do you already have a plan for when you review my contribution? 🙂 

Best,
Tassilo

**Review by korthout — CHANGES_REQUESTED** [review]
Thanks @tasso94 for the improvements. Sorry for taking so long. I want to give this the attention it deserves, but it was hard to find the time. Now, I finally had a chance to give it a spin, and I'm happy to report that it works! It's really cool to see it in action! 🚀 

- Original: https://github.com/korthout/backport-action-test/pull/365#issuecomment-1962945749
- Backport: https://github.com/backport-action/backport-action-test/pull/2

Some notes:
- ❌ There's a small mistake in one of the logs. It refers to `origin`, but actually pushed to `target`. This should be changed before merging. See below.
- 💭 I've tried to think a bit more on our usage of the word `remote` to mean the other repository. [GitHub uses `upstream`](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/about-forks) as the name for the original repository from the perspective of a forked repository. In that same analogy, what we're currently calling `remote` is actually the `downstream` repository. Changes from the upstream repository are backported to the downstream repository. However, I worry that `downstream` is too unfamiliar to be used here. So, perhaps remote is still the best wording. I'd love your input on this.
- 📝 since the default `GITHUB_TOKEN` does not have write access to the other repository, it's mandatory to use a [PAT](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/managing-your-personal-access-tokens) for the `token` input of the checkout action and the `github_token` input of the backport-action itself. We should document this alongside these inputs.

```
Push branch to origin
git push --set-upstream target backport-365-to-case5-backport-target
```

In addition to my ramblings on naming things and the small logging mistake, I only have a few minor suggestions. Please have a look 🙇 

👍 I'll be happy to merge this with a bit of documentation

**Comment by tasso94** [comment]
Hi @korthout,

Thank you for the review and for bringing up the naming.  🙂 👍 

I prefer `downstream` to `remote` since the latter is super generic, whereas the former is self-explanatory. 

What do you think? Should we move forward with `downstream`? 🤔

**Comment by korthout** [comment]
Let's go for it!

**Comment by korthout** [comment]
@tasso94 is there anything I can do to help get this merged?

File: action.yml
Comment (korthout) [inline action.yml:96]: ❌ This has been added individually by #416, let's remove it here

**Comment by tasso94** [comment]
Hi @korthout,

I was quite busy in the last couple of weeks.
I'll did some final touches and incorporated your latest feedback. I'll have a final look Tomorrow. Thanks for your patience! 🙂 👍 

Best,
Tassilo

**Comment by tasso94** [comment]
Hi @korthout,

Friendly reminder to review my latest changes 🙂 👍 

Best,
Tassilo

**Review by korthout — APPROVED** [review]
🙇 Sorry for taking so long. I was in the process of moving internationally. I'm trying to catch up again now.

❤️ Thanks for the amazing work on this! I love the solutions you found and I'm happy with the naming.

I've taken the liberty to adjust some last small things, but nothing major.

Let's merge this! 🚀 Please try it out in your repository and let me know if you run into any issues. In time, we can promote the inputs from their experimental status.

LGTM 👍

**Comment by korthout** [comment]
To try this out you can either use:

- `korthout/backport@main`
- `korthout/backport@v2.next-preview`

File: src/backport.ts
Comment (mvayngrib) [inline src/backport.ts:75]: nit: if this doesn't need to be used externally, better make it private

File: src/backport.ts
Comment (mvayngrib) [inline src/backport.ts:79]: same, could be private

**Review by mvayngrib — COMMENTED** [review]
left a few small suggestions

**Comment by mvayngrib** [comment]
thanks @korthout, i haven't yet, but will share feedback when i do 👍

**Pull request merged**
